### PR TITLE
MACRO

### PR DESCRIPTION
**Pull request by PrezesMajkel**

### CODE REVIEW

File: src/main/java/edu/kis/powp/jobs2d/drivers/MacroDriver.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public void setPosition(int x, int y) {
+        macro.add((new SetPositionCommand(x, y)));
+    }
+
+    public List<DriverCommand> getMacro(){
+        return macro;
+    }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/MacroDriver.java]: Please use "deep copy" feature to return a copy of macro, so client:
-  has full control over the returned command 
- -and cannot modify the state of MacroDriver

File: src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import java.awt.event.ActionListener;
+import java.util.List;
+
+public class SelectLoadMacroOptionListener implements ActionListener {
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        List<DriverCommand> commands = MacroFeature.getMacroDriver().getMacro();
+
+        DriverCommandManager manager = CommandsFeature.getDriverCommandManager();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java]: Please remove dependency with CommandFeature (you can store driverCommandManager as a field).

File: src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.util.List;
+
+public class SelectLoadMacroOptionListener implements ActionListener {
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        List<DriverCommand> commands = MacroFeature.getMacroDriver().getMacro();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java]: Please remove dependency with MacroFeature (you can store macroDriver as a field).

File: src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java [hunk 4]
```diff
@@ -70,6 +70,9 @@ private static void setupDrivers(Application application) {
 
 		driver = new LineDriverAdapter(drawerController, LineFactory.getSpecialLine(), "special");
 		DriverFeature.addDriver("Special line Simulator", driver);
+
+		DriverFeature.addDriver("Macro recording", MacroFeature.getMacroDriver());
+
```
Comment (kowallus) [inline src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java]: macro should be recorded in background as actual drivers work. Now we cannot record and see the effects of operations on the driver. There are two options:
- you can redesign MacroDriver to decorator pattern
- you can take advantage of driver composite feature and addDriver through AdditionalDriverFeature.

**Review by kowallus — CHANGES_REQUESTED** [review]
Almost fine, but please improve your solution.

**Comment by kowallus** [comment]
Please also make sure that there are no conflicts with main repo

File: src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java (hunk 4 above)
Comment (PrezesMajkel) [reply src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java]: Ta funkcjonalność istnieje, mamy dwie opcje skorzystania z MacroDriver, 
jedna to uruchomienie go jako podstawowy sterownik, co spowoduje nagranie makra z wywoływanych poleceń bez żadnej innej akcji;
druga - wybrać "Macro recording" z  "Additional Driver", tak długo ja ta opcja będzie aktywna makro będzie nagrywane, niezależnie od wybranego głównego sterownika

File: src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java [hunk 5]
```diff
@@ -1,21 +1,19 @@
...
 import edu.kis.powp.jobs2d.features.CommandsFeature;
 import edu.kis.powp.jobs2d.features.MacroFeature;
 
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
-import java.util.List;
 
 public class SelectLoadMacroOptionListener implements ActionListener {
+    MacroDriver macroDriver = MacroFeature.getMacroDriver();
+    DriverCommandManager manager = CommandsFeature.getDriverCommandManager();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java]: you haven't removed the dependencies (imports are still there). Please use dependency injection (by constructor)

**Review by kowallus — CHANGES_REQUESTED** [review]
I agree with using Macro as an "Additional Driver". 
Unfortunately, other issues are not fixed yet.

File: src/main/java/edu/kis/powp/jobs2d/events/SelectClearMacroListener.java [hunk 6]
```diff
@@ -3,11 +3,17 @@
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
 
-import edu.kis.powp.jobs2d.features.MacroFeature;
+import edu.kis.powp.jobs2d.drivers.MacroDriver;
 
 public class SelectClearMacroListener implements ActionListener {
+    MacroDriver macroDriver;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/events/SelectClearMacroListener.java]: should be private!

File: src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java [hunk 7]
```diff
@@ -2,15 +2,18 @@
...
-import edu.kis.powp.jobs2d.features.MacroFeature;
 
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
 
 public class SelectLoadMacroOptionListener implements ActionListener {
-    MacroDriver macroDriver = MacroFeature.getMacroDriver();
-    DriverCommandManager manager = CommandsFeature.getDriverCommandManager();
+    MacroDriver macroDriver;
+    DriverCommandManager manager;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/events/SelectLoadMacroOptionListener.java]: same issue

File: src/main/java/edu/kis/powp/jobs2d/drivers/MacroDriver.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    public void setPosition(int x, int y) {
+        macro.add((new SetPositionCommand(x, y)));
+    }
+
+    public CompoundCommand getMacro(){
+        return new CompoundCommand(macro, "Macro");
+    }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/MacroDriver.java:27]: I have commented this previously, but the comment seemed to vanish (maybe it wasn't saved). This is not a deep copy. Check and apply one of the features introduced by another group:
- A visitor for a deep copy of a complex command
- A deep copy of a complex command

**Review by kowallus — CHANGES_REQUESTED** [review]
one previous comment was missing and some minor issues. Please fix.

**Pull request merged**